File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# iterative way without using any stack.


class Solution:
    def preorderTraversal(self, root):
        res = []
        if root is None:
            return res

        curr = root
        while curr is not None:
            res.append(curr.val)
            if curr.left is None:
                curr = curr.right
            else:
                if curr.right is not None:
                    pre = curr.left
                    while pre.right is not None:
                        pre = pre.right
                    pre.right = curr.right
                    curr.right = None
                curr = curr.left

File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution3.js

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// iterative approach without using any stacks.
let preorderTraversal = function (root) {
  let res = []
  if (root === null) return res
  let curr = root
  let pre = null
  while (curr !== null) {
    res.push(curr.val)
    // if curr.left == null, navigate its right child.
    if (curr.left == null) { curr = curr.right } else {
      // if curr.right is not null, make it as the right child of the rightmost child of curr's left child.
      if (curr.right !== null) {
        pre = curr.left
        while (pre.right !== null) { pre = pre.right }
        pre.right = curr.right
        curr.right = null
      }
      // navigate to curr's left child.
      curr = curr.left
    }
  }
}
